beamline.rst

```diff
@@ -1,212 +1,379 @@
-Beamline API
-~~~~~~~~~~~~
+General API
+===========
 
-Functionality that incorporates several different instruments or that does
-not necessarily belong to one particular instrument.
+  Functionality (Procedures, Actuators) used generally across the interface.
 
+    # NB I propose using the classes here globally across the interface.
+    #    We need to decide if we want a single namespace for e.g. procedures,
+    #    or we want to have nested dictionaries to divide them by component.
 
 
-**API Functions**
+
+API Functions
+-------------
 
 .. code:: python
 
-   class ProcedureData(NamedTuple):
-    """
-    Describes a "generic" beamline procedure, procedures that are not
-    necassarily present on all beamlines.
-
-    where:
-
-    cmd_id: the string identifying the command, i.e. 'QUICK_REALIGN'
-    running: True if the procedure is running otherwise False
-    args: arguments to pass to the procedure
-    display_name: the name to display in the user interface i.e. "Quick Realign"
-    tool_tip: the tool tip
-    messages: any initial messages to display
-    """
-        cmd_id: str
+   from typing import *
+
+   from enum import Enum
+
+   **# PROCEDURES**
+
+    class ProcedureData(NamedTuple):
+        """
+        Describes a "generic" beamline procedure, procedures that are not
+        necessarily present on all beamlines.
+
+        where:
+
+        name: the string identifying the command, i.e. 'QUICK_REALIGN'
         state: CmdState
-        args: tuple
-        display_name: str
-        tool_tip: str
-        messages: list
+        args: tuple of arguments to pass to the procedure
+        kwargs: dict[str:Any] of keyword arguments with their defaults
+        display_name: the name to display in the user interface i.e. "Quick Realign"
+        tool_tip: the tool tip
+        messages: any initial messages to display
+        """
+
+        name: str   # NB This assumes that only one instance of a command can run at any time
```

Comment:
Not really that depends on the logic that executes the procedure this is just procedure data, however that semantics gets a bit more complicated if we need to consider multiple procedures of the same type at the same time. We would ideally (but not necessarily) need a procedure id which is tied to a particular instance of a procedure to handle the different run time scenarios 

That aside, we don't really have that need for the moment since this corresponds to procedures involving hardware, like quick realign, we definitely do not want to run several of those at the same time ...

Id even say that its desirable that the logic that executes the procedure prevents several procedures of the same type to run at the same time

Reply:
That is the question, I guess. Do we need a procedure ID separate from the 'understandable' name? I guess the answer is no (for now).